src/transform/src/attribute/non_negative.rs

```diff
@@ -0,0 +1,221 @@
+                    // 1. Reset the entry in the env map with the shadowed value.
+                    self.env.insert(id, val);
+                    // 2. Create a NoOp task indicating that there is nothing more to be done.
+                    EnvTask::NoOp
+                }
+                // An Remove task indicates that we are about to leave the Let binding body
+                // and the id of the Let parent did not shadow another `id` in the environment.
+                EnvTask::Remove(id) => {
+                    // Before moving to the post_visit of the enclosing Let parent, do as follows:
+                    // 1. Remove the value assciated with the given `id` from the environment.
+                    self.env.remove(&id);
+                    // 2. Create a NoOp task indicating that there is nothing more to be done.
+                    EnvTask::NoOp
+                }
+                // A NoOp task indicates that we don't need to do anyting.
+                EnvTask::NoOp => EnvTask::NoOp,
+            };
+            // Advance the state machine.
+            self.env_tasks.push(env_task);
+        };
+    }
+}
+
+impl NonNegative {
+    fn derive(&mut self, expr: &MirRelationExpr) {
+        use MirRelationExpr::*;
+        let n = self.results.len();
+        match expr {
+            Constant { rows, .. } => {
+                if let Ok(rows) = rows {
+                    let has_negative_rows = rows.iter().any(|(_data, diff)| diff < &0);
+                    self.results.push(has_negative_rows);
+                } else {
+                    self.results.push(true); // constant errors are considered "non-negative"
+                }
+            }
+            Get { id, .. } => match id {
+                Id::Local(id) => match self.env.get(id) {
+                    Some(value) => self.results.push(value.clone()),
+                    None => self.results.push(false),
+                },
+                Id::Global(_) => {
+                    self.results.push(true);
+                }
+            },
+            Let {
+                value: _, body: _, ..
+            } => {
+                let body = self.results[n - 1];
+                self.results.push(body);
+            }
+            Project { input: _, .. } => {
+                let input = self.results[n - 1];
+                self.results.push(input);
+            }
+            Map { input: _, .. } => {
+                let input = self.results[n - 1];
+                self.results.push(input);
+            }
+            FlatMap { input: _, .. } => {
+                let input = self.results[n - 1];
+                self.results.push(input);
+            }
+            Filter { input: _, .. } => {
+                let input = self.results[n - 1];
+                self.results.push(input);
+            }
+            Join { inputs, .. } => {
+                let inputs = &self.results[n - inputs.len()..];
```

Comment:
Wouldn't these also need the `subtree_size` stuff? (and the same for `union` below)

Reply:
It seems that this is really bugging my mind...

Reply:
Can you re-post this as a top-level comment. I have implemented the post-order based approach correctly I think, but it would be good to know that an alternative exists.